File: T-Bank/Algorithms-and-Data-Structures-2025/9sprint-SegmentTree/contest/F-Противник_слаб/solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <map>

using namespace std;

class FenwickTree {
private:
    int size;
    vector<int> tree;

public:
    FenwickTree(int size) : size(size), tree(size + 2, 0) {}

    void update(int index, int delta = 1) {
        while (index <= size) {
            tree[index] += delta;
            index += index & -index;
        }
    }

    int query(int index) {
        int res = 0;
        while (index > 0) {
            res += tree[index];
            index -= index & -index;
        }
        return res;
    }
};


void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    */
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; ++i) {
        cin >> nums[i];
    }
    vector<int> sorted_nums = nums;
    sort(sorted_nums.begin(), sorted_nums.end());
    sorted_nums.erase(unique(sorted_nums.begin(), sorted_nums.end()), sorted_nums.end());
    map<int, int> compressed;
    for (int i = 0; i < sorted_nums.size(); ++i) {
        compressed[sorted_nums[i]] = i + 1;
    }
    int max_compressed = sorted_nums.size();
    vector<int> nums_compressed(n);
    for (int i = 0; i < n; ++i) {
        nums_compressed[i] = compressed[nums[i]];
    }
    vector<int> left(n);
    FenwickTree ft_left(max_compressed);
    for (int i = 0; i < n; ++i) {
        left[i] = i - ft_left.query(nums_compressed[i]);
        ft_left.update(nums_compressed[i]);
    }
    vector<int> right(n);
    FenwickTree ft_right(max_compressed);
    for (int i = n - 1; i >= 0; --i) {
        right[i] = ft_right.query(nums_compressed[i] - 1);
        ft_right.update(nums_compressed[i]);
    }
    long long weakness = 0;
    for (int i = 0; i < n; ++i) {
        weakness += static_cast<long long>(left[i]) * right[i];
    }
    cout << weakness << '\n';
}


int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/9sprint-SegmentTree/contest/F-Противник_слаб/solution.py
import sys
import bisect


class FenwickTree:
    def __init__(self, size: int) -> None:
        self.size: int = size
        self.tree: list[int] = [0] * (self.size + 2)
    
    def update(self, index: int, delta: int = 1) -> None:
        while index <= self.size:
            self.tree[index] += delta
            index += index & -index
    
    def query(self, index: int) -> None:
        res: int = 0
        while index > 0:
            res += self.tree[index]
            index -= index & -index
        return res


def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline())
    nums: list[int] = list(map(int, sys.stdin.readline().split()))
    sorted_nums: list[int] = sorted(set(nums))
    compressed: dict[int, int] = {v: i + 1 for i, v in enumerate(sorted_nums)}
    max_compressed: int = len(sorted_nums)
    nums_compressed: list[int] = [compressed[v] for v in nums]
    left: list[int] = [0] * n
    ft_left: FenwickTree = FenwickTree(max_compressed)
    for i in range(n):
        left[i] = i - ft_left.query(nums_compressed[i])
        ft_left.update(nums_compressed[i])
    
    right: list[int] = [0] * n
    ft_right: FenwickTree = FenwickTree(max_compressed)
    for i in range(n - 1, -1, -1):
        right[i] = ft_right.query(nums_compressed[i] - 1)
        ft_right.update(nums_compressed[i])
    
    weakness: int = sum(left[i] * right[i] for i in range(n))
    sys.stdout.write('{}\n'.format(weakness))

if __name__ == "__main__":
    solution()